torch/csrc/api/include/torch/data/init.h

```diff
@@ -0,0 +1,160 @@
+template <typename DataType>
+class PyChunkDataReader : public ChunkDataReader<DataType> {
+ public:
+  // Inherit the constructors
+  using ChunkDataReader<DataType>::ChunkDataReader;
+
+  // Trampoline (need one for each virtual function)
+  typename ChunkDataReader<DataType>::ChunkType read_chunk(
+      size_t chunk_index) override {
+    PYBIND11_OVERLOAD_PURE(
+        typename ChunkDataReader<DataType>::ChunkType,
+        ChunkDataReader<DataType>,
+        read_chunk,
+        chunk_index);
+  }
+  size_t chunk_count() override {
+    PYBIND11_OVERLOAD_PURE(size_t, ChunkDataReader<DataType>, chunk_count);
+  }
+  void reset() override {
+    PYBIND11_OVERLOAD_PURE(void, ChunkDataReader<DataType>, reset);
+  }
+};
+
+/* Generates Python bindings for a specific ChunkDataReader implementation
+ *
+ * A reference for the binded class is returned so that its constructors
+ * are manually binded by the user. This is necessary as the constructor
+ * signature are implementation specific
+ */
+template <typename SpecificChunkDataReader, typename... Args>
+py::class_<SpecificChunkDataReader, Args...>  bind_chunkdatareader(
+    py::module& m,
+    const std::string& typestr) {
+  // SpecificChunkDataReader
+  return (py::class_<SpecificChunkDataReader, Args...>(
+             m,
+             typestr.c_str(),
+             "Performs data chunking and reading of entire chunks")
+      .def(
+          "read_chunk",
+          &SpecificChunkDataReader::read_chunk,
+          "Returns data from a chunk file",
+          py::arg("chunk_index"),
+          py::return_value_policy::take_ownership)
+      .def(
+          "chunk_count",
+          &SpecificChunkDataReader::chunk_count,
+          "Returns the number of chunks")
+      .def("reset", &SpecificChunkDataReader::reset, "Not used"));
+}
+
+/* Generates Python bindings for a specific ChunkDataset implementation */
+template <typename SpecificChunkDataReader>
+void bind_chunkdataset(py::module& m, const std::string& typestr) {
+  // SpecificChunkDataset
+  using SpecificChunkDataset = ChunkDataset<
+      SpecificChunkDataReader,
+      samplers::SamplerWrapper,
+      samplers::SamplerWrapper>;
+  py::class_<SpecificChunkDataset>(
+      m,
+      typestr.c_str(),
+      "A stateful dataset that supports hierarchical sampling and prefetching of entire chunks."
+      " Unlike regular `Dataset`, `ChunkDataset` requires two samplers to operate and keep internal state:"
+      " 1) A chunk sampler selects which chunk to load next; and"
+      " 2) An example sampler determines the order of `Example`s that are returned in each `get_batch` call.")
+      .def(
+          py::init([](SpecificChunkDataReader chunk_reader,
+                      ChunkDatasetOptions options,
+                      bool shuffle_chunks,
+                      bool shuffle_samples,
+                      size_t num_replicas,
+                      size_t rank,
```

Comment:
I just wanted to double check that you're aware that this means that you will never be able to use other samplers with this kind of dataset from Python without breaking backwards compatibility?